Implement the form owner concept

canova opened the pull request:
<!-- Please describe your changes on the following line: -->
Implemention of the core logic for form owners. It's rebased version of #6613, addressed most of the comments and fixed some of the code. I needed to update the code by hand instead of rebasing old main commit because of the massive merge conflicts over the years.

The html5ever PR is here: https://github.com/servo/html5ever/pull/249

Currently I'm using my html5ever fork to run the tests. I'll undo the changes in Cargo.toml and Cargo.lock once it gets merged.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #3553 (github issue number if applicable).

<!-- Either: -->
- [X] There are tests for these changes

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/15283)
<!-- Reviewable:end -->

mukilan committed:
Update web-platform-tests expected data

mukilan committed:
Add tests for form owner concept

mukilan committed:
Implement the form owner concept

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/Cargo.toml, components/script/dom/htmlselectelement.rs, components/script/dom/document.rs, components/script/dom/htmllegendelement.rs, components/script/dom/element.rs, components/script/dom/htmlfieldsetelement.rs, components/script/dom/servoparser/html.rs, components/script/dom/virtualmethods.rs, components/script/dom/htmlobjectelement.rs, components/script/dom/htmllabelelement.rs, components/script/dom/htmltextareaelement.rs, components/script/dom/htmloutputelement.rs, components/script/dom/htmlimageelement.rs, components/script/dom/htmlinputelement.rs, components/script/dom/htmlbuttonelement.rs, components/script/dom/node.rs, components/script/dom/htmlformelement.rs
 * @kichjang: components/script/Cargo.toml, components/script/dom/htmlselectelement.rs, components/script/dom/document.rs, components/script/dom/htmllegendelement.rs, components/script/dom/element.rs, components/script/dom/htmlfieldsetelement.rs, components/script/dom/servoparser/html.rs, components/script/dom/virtualmethods.rs, components/script/dom/htmlobjectelement.rs, components/script/dom/htmllabelelement.rs, components/script/dom/htmltextareaelement.rs, components/script/dom/htmloutputelement.rs, components/script/dom/htmlimageelement.rs, components/script/dom/htmlinputelement.rs, components/script/dom/htmlbuttonelement.rs, components/script/dom/node.rs, components/script/dom/htmlformelement.rs

canaltinova commented:
r? @jdm 
New added `form_owner_and_table.html` test is failing currently but I'll investigate the problem tomorrow. Other added tests are currently passing. I need to also run a try build to see if I broke something.

canaltinova commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit ef9cac0 with merge bbc4f64...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/5434)

canaltinova commented:
Running builds: 
http://build.servo.org/builders/linux-rel-css/builds/2115

jdm commented:
```
  ▶ Unexpected subtest result in /html/semantics/forms/form-control-infrastructure/form.html:
  └ PASS [expected FAIL] label-in-table-with-control.form

  ▶ Unexpected subtest result in /html/semantics/forms/form-control-infrastructure/form.html:
  └ PASS [expected FAIL] label-in-table-for.form

  ▶ Unexpected subtest result in /html/semantics/forms/form-control-infrastructure/form.html:
  │ FAIL [expected PASS] label.form
  │   → assert_equals: label.form expected null but got Element node <form id="form">
<p><button id="button">button</button>
<...
  │ 
  │ testLabel/<@http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:81:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ testLabel@http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:78:3
  └ @http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:85:1

  ▶ Unexpected subtest result in /html/semantics/forms/form-control-infrastructure/form.html:
  │ FAIL [expected PASS] label-in-table.form
  │   → assert_equals: label.form expected null but got Element node <form id="form3"></form>
  │ 
  │ testLabel/<@http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:81:5
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  │ testLabel@http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:78:3
  └ @http://web-platform.test:8000/html/semantics/forms/form-control-infrastructure/form.html:94:1

  ▶ CRASH [expected OK] /html/semantics/forms/resetting-a-form/reset-form.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ not yet implemented (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(0), index: PipelineIndex(1) }, at /Users/servo/buildbot/slave/mac-rel-wpt1/build/components/script/dom/htmlformelement.rs:654)
  │ stack backtrace:
  │    0:        0x10534bc6e - backtrace::backtrace::trace::h05b30eb9c9ea9aaa
  │    1:        0x10534bf7c - backtrace::capture::Backtrace::new::hd6fb6e83f9059b0d
  │    2:        0x1041d00f4 - servo::main::{{closure}}::hfe009f9dcda553cc
  │    3:        0x105d05564 - std::panicking::rust_panic_with_hook::h33761bada49f3713
  │    4:        0x1043f3534 - std::panicking::begin_panic::hf16f59ea153d03ea
  │    5:        0x1047440f0 - script::dom::htmlformelement::HTMLFormElement::reset::h128594eee2b14ab3
  │    6:        0x10444e33b - std::panicking::try::do_call::h4300f1ebd1e5b90e
  │    7:        0x105d0647a - __rust_maybe_catch_panic
  │    8:        0x1049ce020 - script::dom::bindings::codegen::Bindings::HTMLFormElementBinding::HTMLFormElementBinding::reset::he27a56e267712168
  │    9:        0x104b57e68 - CallJitMethodOp
  │   10:        0x104674113 - script::dom::bindings::utils::generic_call::h97caee99528742d7
  │   11:        0x104fc1aee - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │   12:        0x104fbb30c - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   13:        0x104fae9ca - 2js9RunScriptEP9JSContextRNS_8RunState
  │   14:        0x104fc1e42 - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │   15:        0x104fc1fa5 - 2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_E
  │   16:        0x104eee4d0 - 2js9fun_applyEP9JSContextjPN2JS5Value
  │   17:        0x104fc1aee - 2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstruct
  │   18:        0x104fbb30c - _ZL9InterpretP9JSContextRN2js8RunStateE
  │   19:        0x104fae9ca - 2js9RunScriptEP9JSContextRNS_8RunState
  │   20:        0x104fc2981 - _ZN2js13ExecuteKernelEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectRKNS2_5ValueENS_16AbstractFramePtrEPS9_
  │   21:        0x104fc2ad7 - 2js7ExecuteEP9JSContextN2JS6HandleIP8JSScriptEER8JSObjectPNS2_5Value
  │   22:        0x104e90a88 - _ZL8EvaluateP9JSContextN2JS6HandleIP8JSObjectEENS2_IPN2js11StaticScopeEEERKNS1_22ReadOnlyCompileOptionsERNS1_18SourceBufferHolderENS1_13MutableHandleINS1_5ValueEEE
  │   23:        0x104e90c45 - 2JS8EvaluateEP9JSContextRKNS_22ReadOnlyCompileOptionsEPKDsmNS_13MutableHandleINS_5ValueEE
  │   24:        0x104719739 - script::dom::globalscope::GlobalScope::evaluate_script_on_global_with_result::ha18c0a0349b7083e
  │   25:        0x104776193 - script::dom::htmlscriptelement::HTMLScriptElement::execute::h499aa81fb515cfb7
  │   26:        0x10477439b - script::dom::htmlscriptelement::HTMLScriptElement::prepare::hb1bdd2f230f2fb89
  │   27:        0x1047c8dfc - script::dom::servoparser::ServoParser::do_parse_sync::h34666ed83b35150e
  │   28:        0x1047c8b66 - script::dom::servoparser::ServoParser::parse_sync::h101a970af8f26898
  │   29:        0x1047c76b6 - script::dom::servoparser::ServoParser::resume_with_pending_parsing_blocking_script::ha976fdf55ff1ce94
  │   30:        0x1046d3c7e - script::dom::document::Document::process_pending_parsing_blocking_script::hfa05777ecaf76d70
  │   31:        0x1046d3885 - script::dom::document::Document::pending_parsing_blocking_script_loaded::h9664964978802b97
  │   32:        0x1047732bf - <script::dom::htmlscriptelement::ScriptContext as net_traits::FetchResponseListener>::process_response_eof::h0ad6366679da3973
  │   33:        0x104846abd - <script::network_listener::ListenerRunnable<A, Listener> as script::script_thread::Runnable>::handler::h73af4d13e3698d15
  │   34:        0x10484e882 - <script::script_thread::CancellableRunnable<T> as script::script_thread::Runnable>::handler::hee626778b24667da
  │   35:        0x1048684c1 - script::script_thread::ScriptThread::handle_msg_from_script::h599af8ff8a2bc0fd
  │   36:        0x1048631d5 - script::script_thread::ScriptThread::handle_msgs::{{closure}}::h6d4abf84d28ef2c3
  │   37:        0x10485e81b - script::script_thread::ScriptThread::handle_msgs::h0625f27481ecd3b8
  │   38:        0x10451cdd7 - std::panicking::try::do_call::he05758f7896ded07
  │   39:        0x105d0647a - __rust_maybe_catch_panic
  │   40:        0x1045ab0ba - <F as alloc::boxed::FnBox<A>>::call_box::heeb5f37ee44a8ef5
  │   41:        0x105d044e4 - std::sys::imp::thread::Thread::new::thread_start::he1b44499d44cc4a4
  │   42:     0x7fff8afbb059 - _pthread_body
  │   43:     0x7fff8afbafd6 - _pthread_start
  │ ERROR:servo: not yet implemented
  └ Pipeline failed in hard-fail mode.  Crashing!

  ▶ Unexpected subtest result in /html/semantics/forms/the-legend-element/legend-form.html:
  │ FAIL [expected PASS] Check if legend.form returns its parent when it's inside a fieldset
  │   → assert_not_equals: got disallowed value null
  │ 
  │ @http://web-platform.test:8000/html/semantics/forms/the-legend-element/legend-form.html:26:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:501:9
  └ @http://web-platform.test:8000/html/semantics/forms/the-legend-element/legend-form.html:24:1
```

canova committed:
Fix failing tests

canova committed:
Use one form_attribute_mutated method instead of set and remoted

canova committed:
Fix test by not calling reset_form_owner() in unbind_form_control_from_tree when it shouldn't

canova committed:
Exclude label element from form_attribute.html

Because label has a different form owner concept. It should reflect its control's form owner. When it doesn't have a control it should print null.

canaltinova commented:
@jdm I think it's ready to review now :) Failing tests are passing and removed some fail expectations. I want to test it one more time to see if everything is ok.
@bors-servo try

bors-servo commented:
:hourglass: Trying commit c60dad5 with merge 22718c8...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2174)

canova committed:
Update test expectations

jdm commented:
I have only skimmed the changes so far, but I would like to find a way to reduce all of the duplicate implementations of FormControl, bind_to_tree, etc. What if we added a `as_form_control` helper (like `as_maybe_activatable`) that we could use from `Element::bind_to_tree`?

canaltinova commented:
I suppose we can do that, I'll try this.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15405) made this pull request unmergeable. Please resolve the merge conflicts.

canova committed:
Move control's tree bind/unbind calls to Element's tree bind/unbind methods

canaltinova commented:
I moved the `bind_form_control_to_tree/unbind_form_control_from_tree` calls to Element's bind/unbind function. This reduced the duplication a bit.

There are still FormControl's methods to move. But they look like they can't removed easily(AFAIK). Because `form_owner` and `set_form_owner` methods access directly to `form_owner` field of the struct. And we don't know if this struct has that field or not. (For example `HTMLLabelElement` doesn't need to store a form_owner field.) Also `to_element` was predefined inside FormControl before, but we needed to remove `DerivedFrom<Element>` from FormControl trait to be able to use `as_maybe_form_control` helper method. And we can't upcast like this `self.upcast::<Element>()` if trait is not derived from Element. And not every form control has form attribute, so I can't generalize `attribute_mutated` either.

jdm commented on tests/wpt/mozilla/tests/mozilla/form_attribute.html:
```diff
@@ -0,0 +1,234 @@
+<!DOCTYPE html>
```
All of these tests belong in tests/wpt/web-platform-tests/html/semantics/forms/form-control-infrastructure/ since they're testing behaviour that is described in the specification.

jdm commented on components/script/dom/htmlformelement.rs:
```diff
@@ -651,7 +652,7 @@ impl HTMLFormElement {
                 NodeTypeId::Element(ElementTypeId::HTMLElement(HTMLElementTypeId::HTMLOutputElement)) => {
                     // Unimplemented
-                    unimplemented!()
+                    {}
```
We should just be able to remove the {} here.

jdm commented on components/script/dom/htmllabelelement.rs:
```diff
@@ -174,11 +172,18 @@ impl HTMLLabelElement {
 impl FormControl for HTMLLabelElement {
     fn form_owner(&self) -> Option<Root<HTMLFormElement>> {
-        self.form_owner.get()
+        if let Some(e) = self.GetControl() {
```
```rust
self.GetControl().and_then(|e| {
    let form_control = e.upcast::<Element>().as_maybe_form_control()
                                .expect("Element must be a form control");
    form_control.form_owner()
})
```

jdm commented on components/script/dom/htmlformelement.rs:
```diff
@@ -974,7 +974,15 @@ pub trait FormControl: DomObject {
+        // FIXME: We should call `form_attribute_mutated` method here,
+        // but also we don't want to call reset_form_owner() method in
+        // `form_attribute_mutated`. We should fix this.
+        // self.form_attribute_mutated(AttributeMutation::Removed);
```
Let's split form_attribute_mutated into `register_if_necessary`, `unregister_if_necessary`, and `form_attribute_mutated`. `form_attribute_mutated` can call the other two methods based on the mutation, and code like `bind_form_control_to_tree` can call those methods and reset_form_owner as needed.

jdm reviewed (changes requested):
Thank you so much for tackling this! I've left a bunch of comments, but I appreciate the work you did to update the original PR and clean up some of the code duplication.

canaltinova commented on components/script/dom/htmlformelement.rs:
```diff
@@ -646,14 +651,28 @@ impl HTMLFormElement {
                 NodeTypeId::Element(ElementTypeId::HTMLElement(HTMLElementTypeId::HTMLOutputElement)) => {
                     // Unimplemented
-                    {}
+                    unimplemented!()
```
Yeah, at first I thought this can panic here but later I realized it's breaking some tests. Removed it afterwards.

canova committed:
Address review comments

canaltinova commented on components/script/dom/htmlformelement.rs:
```diff
@@ -843,24 +862,129 @@ impl<'a> FormSubmitter<'a> {
+            }
+        }
+
+        let new_owner = if self.is_reassociatable() && has_form_id && node.is_in_doc() {
```
self isn't `Derived<Element>` in this trait. Because of it we can't call `self.is_connected()` directly. Changed it with `elem.is_connected()` (elem is `self.to_element()`)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15584) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15588) made this pull request unmergeable. Please resolve the merge conflicts.

canaltinova commented:
@jdm Sorry couldn't look at this PR recently. I've addressed most of your comments and rebased.
> Why is this change necessary? Do we have a test that covers it?

I've looked at the algorithm and the spec but couldn't find the reason. It was like this in the previous PR. Saw your comment about this line and thought it should be necessary:
> A comment explaining why it's important to perform these operations separately would help. I had to think about it for a minute.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15715) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #15752) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@jdm Stealing this from you, I'm sure it will make you *insanely* angry to see people steal reviews from your queue.

nox commented:
Superseded by #15938, thanks @mukilan and @canaltinova.
